src/globus_sdk/services/auth/consents.py

```diff
@@ -0,0 +1,311 @@
+        for node in self.nodes:
+            if len(node.dependency_path) > 1:
+                parent_id = node.dependency_path[-2]
+                try:
+                    edges[parent_id].add(node.id)
+                except KeyError as e:
+                    raise ConsentParseError(
+                        f"Failed to compute forest edges. Missing parent node: {e}. "
+                        f"Consents: {self.nodes}."
+                    ) from e
+        return edges
+
+    def _build_trees(self) -> list[ConsentTree]:
+        """
+        Build out the list of trees in the forest.
+
+        A distinct tree is built out for each "root nodes" (nodes with no parents).
+        """
+
+        # A node with dependency path length 1 has no parents, so it is a root.
+        roots = [node for node in self.nodes if len(node.dependency_path) == 1]
+        return [ConsentTree(root.id, self) for root in roots]
+
+    def get_node(self, consent_id: int) -> Consent:
+        return self._node_by_id[consent_id]
+
+    def meets_scope_requirements(self, scopes: Scope | str | list[Scope | str]) -> bool:
+        """
+        Check whether this consent meets one or more scope requirements.
+
+        A consent forest meets a particular scope requirement if any consent tree
+            inside the forest meets the scope requirements.
+
+        :param scopes: A single scope, a list of scopes, or a scope string to check
+           against the forest.
+        :returns: True if all scope requirements are met, False otherwise.
+        """
+        for scope in _normalize_scope_types(scopes):
+            if not any(tree.meets_scope_requirements(scope) for tree in self.trees):
+                return False
+        return True
+
+
+class ConsentTree:
+    """
+    A tree of Consent nodes with edges modeling the dependency relationships between
+        them.
+    """
+
+    def __init__(self, root_id: int, forest: ConsentForest):
+        self.root = forest.get_node(root_id)
+        self.nodes = [self.root]
+        self._node_by_id = {root_id: self.root}
+        self.edges: dict[int, set[int]] = {}
+
+        self._populate_connected_nodes_and_edges(forest)
+
+    def _populate_connected_nodes_and_edges(self, forest: ConsentForest) -> None:
+        """
+        Populate the nodes and edges of the tree by traversing the forest.
+
+        Nodes/Edges are included in the tree iff they are reachable from the root.
+        """
+        nodes_to_evaluate = {self.root.id}
+        while nodes_to_evaluate:
+            consent_id = nodes_to_evaluate.pop()
+            consent = forest.get_node(consent_id)
+            self.edges[consent.id] = forest.edges[consent.id]
+
+            for child_id in self.edges[consent.id]:
+                if child_id not in self._node_by_id:
+                    self.nodes.append(forest.get_node(child_id))
+                    self._node_by_id[child_id] = forest.get_node(child_id)
```

Comment:
Trees only copy references to the data (consent objects).

We copy edges as those are just maps of ints -> sets of ints. Is that the data you're concerned with copying? If so, can you help me understand why that gives you pause?

Reply:
It's not so much about being concerned about some negative runtime impact -- it's a small body of data to copy -- but rather that I actually think the model is clearer with "trees are a view over the forest".

Since there's no possibility that the data deviates between the two, I don't think that the copy is helpful to a maintainer here understanding what's happening.

Reply:
I disagree with you here.

As a general rule, I find pre-materialized objects a lot easier to conceptualize than views. Data objects aren't "a JIT filter applied to some larger set of data", they're just data objects that developers should already be familiar with sitting available for introspection. Internal components model 1-1 onto the external interfaces defined by the class.

I think about it the same way for a database; I'd use a `view` over a `table` not because a view is easier to understand than a table (I'd argue it's a decent amount harder to understand) but because it can optimize for non-ephemeral storage space despite adding a layer of indirection to reason around.

Reply:
I'm not sure that we're going to see eye-to-eye on this, but I'm also _quite sure_ that it's not worth belaboring the point at present and trying to get agreement.

Whenever you copy data, you introduce the possibility, theoretical or otherwise, that the data diverges. Since there is no meaningful/sensible state in which a tree and forest are in disagreement, I prefer (mildly! we've already spent more time on this topic than is probably warranted) a data model which strictly does not allow for that situation.